users/tests.py

```diff
@@ -0,0 +1,1190 @@
+            with patch.object(self.user_model, 'save') as mocked_save:
+                self.user_model.objects.create_user(email=email, password=password)
+                mocked_save.assert_not_called()
+
+
+class OdiSuperUserTestCase(TestCase):
+    def setUp(self) -> None:
+        self.user_model = get_user_model()
+
+    def test_superuser_complete(self):
+        email = 'superuser@email.com'
+        password = 'password'
+        dummy_user = OdiUser(email=email)
+        dummy_user.set_password(password)
+
+        with patch.object(self.user_model.objects, 'create_user') as mocked_object_manager:
+            with patch.object(self.user_model, 'save') as mocked_save:
+                mocked_object_manager.return_value = dummy_user
+                new_superuser = self.user_model.objects.create_superuser(email=email, password=password)
+
+                self.assertEqual(new_superuser.email, email)
+                self.assertTrue(new_superuser.check_password(password))
+                self.assertTrue(new_superuser.is_staff)
+                self.assertTrue(new_superuser.is_admin)
+                self.assertTrue(new_superuser.is_superuser)
+                mocked_save.assert_called_once()
+
+
+class UtilityTestCase(TestCase):
+    def test_validate_password_when_valid(self):
+        validation_result = utils.validate_password('Pass1234')
+
+        self.assertTrue(validation_result['is_valid'])
+        self.assertIsNone(validation_result['message'])
+
+    def test_validate_password_when_length_is_less(self):
+        error_message = 'Password length must be at least 8 characters'
+        validation_result = utils.validate_password('123')
+
+        self.assertFalse(validation_result['is_valid'])
+        self.assertEqual(validation_result['message'], error_message)
+
+    def test_validate_password_when_no_uppercase(self):
+        error_message = 'Password length must contain at least 1 uppercase character'
+        validation_result = utils.validate_password('abcdefgh')
+
+        self.assertFalse(validation_result['is_valid'])
+        self.assertEqual(validation_result['message'], error_message)
+
+    def test_validate_password_when_no_lowercase(self):
+        error_message = 'Password length must contain at least 1 lowercase character'
+        validation_result = utils.validate_password('A12345678')
+
+        self.assertFalse(validation_result['is_valid'])
+        self.assertEqual(validation_result['message'], error_message)
+
+    def test_validate_password_when_no_number(self):
+        error_message = 'Password length must contain at least 1 number character'
+        validation_result = utils.validate_password('Password')
+
+        self.assertFalse(validation_result['is_valid'])
+        self.assertEqual(validation_result['message'], error_message)
+
+    def test_validate_email_when_valid(self):
+        email = 'valid_email@email.com'
+        self.assertTrue(utils.validate_email(email))
```

Comment:
Hi Indi! I think its better to put the string in the function parameter rather than storing it first

Reply:
Hi Arya! Thank you for your suggestion, I'll also keep that in mind for future implementations.